src/turbomind/models/llama/LlamaLinear.h

```diff
@@ -34,29 +35,69 @@ class LlamaLinear {
+                                  input_data,                                     // B
+                                  weight.input_dims,                              // ldb
+                                  output_data + batch_size * weight.output_dims,  // C
+                                  weight.lora.r);                                 // ldc
+
+            cublas_wrapper_->Gemm(CUBLAS_OP_N,
+                                  CUBLAS_OP_N,
+                                  weight.output_dims,                             // m
+                                  batch_size,                                     // n
+                                  weight.lora.r,                                  // k
+                                  (const T*)weight.lora.b,                        // A
+                                  weight.output_dims,                             // lda
+                                  output_data + batch_size * weight.output_dims,  // B
+                                  weight.lora.r,                                  // ldb
+                                  output_data,                                    // C
+                                  weight.output_dims,                             // ldc
+                                  weight.lora.scale,                              // alpha
+                                  0.0f);                                          // beta
+
+            invokeMask(output_data, lora_mask, batch_size, weight.output_dims, stream_);
+
+            beta      = 1.0f;
+            gemm_type = weight.type == WeightType::kINT4 ? GemmS4F16::kFusedAdd : gemm_type;
+        }
         switch (weight.type) {
             case WeightType::kFP16:
             case WeightType::kFP32:
             case WeightType::kBF16:
-                if (lora_mask != nullptr && weight.lora.r > 0) {
-                    forwardFpLora(output_data, input_data, batch_size, weight, type, lora_mask);
-                }
-                else {
-                    forwardFp(output_data, input_data, batch_size, weight, type);
-                }
+                forwardFp(output_data, input_data, batch_size, weight, gemm_type, alpha, beta);
                 break;
             case WeightType::kINT4:
-                forwardInt4(output_data, input_data, batch_size, weight, type);
+                forwardInt4(output_data, input_data, batch_size, weight, gemm_type);
+                break;
                 break;
             default:
                 FT_CHECK(0);
         }
     }
 
 private:
-    void forwardFp(T* output_data, const T* input_data, int batch_size, const LlamaDenseWeight<T>& weight, Type type)
+    void forwardFp(T*                         output_data,
+                   const T*                   input_data,
+                   int                        batch_size,
+                   const LlamaDenseWeight<T>& weight,
+                   GemmS4F16::Type            type,
```

Comment:
cublas_wrapper doesn't use `type`

Reply:
it is not used by cublas_wrapper

Reply:
I think it's unnecessary to define the parameter "GemmS4F16::Type type" since it is not used.